prog/weaver/http.go

```diff
@@ -0,0 +1,210 @@
+   DirectPeers: {{len .Router.ConnectionMaker.DirectPeers}}
+{{if .IPAM}}\
+
+       Service: ipam
+{{if .IPAM.Entries}}\
+     Consensus: achieved
+{{else if .IPAM.Paxos}}\
+     Consensus: waiting (quorum: {{.IPAM.Paxos.Quorum}}, known: {{.IPAM.Paxos.KnownNodes}})
+{{else}}\
+     Consensus: deferred
+{{end}}\
+         Range: {{.IPAM.Range}}
+ DefaultSubnet: {{.IPAM.DefaultSubnet}}
+{{end}}\
+{{if .DNS}}\
+
+       Service: dns
+        Domain: {{.DNS.Domain}}
+       Address: {{.DNS.Address}}
+           TTL: {{.DNS.TTL}}
+       Entries: {{countDNSEntries .DNS.Entries}}
+{{end}}\
+`)
+
+var peersTemplate = defTemplate("peers", `\
+{{range .Router.Peers}}\
+{{.Name}}({{.NickName}})
+{{range .Connections}}\
+   {{if .Outbound}}->{{else}}<-{{end}} \
+{{printf "%-21v" .Address}} \
+{{$nameNickName := printf "%v(%v)" .Name .NickName}}\
+{{printf "%-32v" $nameNickName}} \
+{{if .Established}}established{{else}}unestablished{{end}}
+{{end}}\
+{{end}}\
+`)
+
+var connectionsTemplate = defTemplate("connectionsTemplate", `\
+{{$directPeers := .Router.ConnectionMaker.DirectPeers}}\
+{{$ourself := .Router.Name}}\
+{{range .Router.Peers}}\
+{{if eq .Name $ourself}}\
+{{range .Connections}}\
+{{if .Outbound}}->{{else}}<-{{end}} \
+{{printf "%-21v" .Address}} \
+{{$nameNickName := printf "%v(%v)" .Name .NickName}}\
+{{printf "%-32v" $nameNickName}} \
+{{$connectionType := connectionType $directPeers .Address}}\
+{{printf "%-10v" $connectionType}} \
+{{if .Established}}established{{else}}unestablished{{end}}
+{{end}}\
+{{end}}\
+{{end}}\
+{{range .Router.ConnectionMaker.Reconnects}}\
+-> \
+{{printf "%-21v" .Address}} \
+                                 \
+{{$connectionType := connectionType $directPeers .Address}}\
+{{printf "%-10v" $connectionType}} \
+{{if .Attempting}}\
+{{if .LastError}}\
+retrying({{.LastError}})
+{{else}}\
+connecting
+{{end}}\
+{{else}}\
+failed({{.LastError}}), retry: {{.TryAfter}}
+{{end}}\
+{{end}}\
+`)
+
+var dnsEntriesTemplate = defTemplate("dnsEntries", `\
+{{$domain := printf ".%v" .DNS.Domain}}\
+{{range .DNS.Entries}}\
+{{if eq .Tombstone 0}}\
+{{$hostname := trimSuffix .Hostname $domain}}\
+{{printf "%-12v" $hostname}} {{printf "%-15v" .Address}} {{printf "%12.12v" .ContainerID}} {{.Origin}}
+{{end}}\
+{{end}}\
+`)
+
+type WeaveStatus struct {
+	Router *weave.Status      `json:"Router,omitempty"`
+	IPAM   *ipam.Status       `json:"IPAM,omitempty"`
+	DNS    *nameserver.Status `json:"DNS,omitempty"`
+}
+
+func NewWeaveStatus(
+	router *weave.Router,
+	allocator *ipam.Allocator,
+	defaultSubnet address.CIDR,
+	ns *nameserver.Nameserver,
+	dnsserver *nameserver.DNSServer) WeaveStatus {
+
+	return WeaveStatus{
+		weave.NewStatus(router),
+		ipam.NewStatus(allocator, defaultSubnet),
+		nameserver.NewStatus(ns, dnsserver)}
+
+}
+
+func HandleHTTP(muxRouter *mux.Router,
+	router *weave.Router,
+	allocator *ipam.Allocator,
+	defaultSubnet address.CIDR,
+	ns *nameserver.Nameserver,
+	dnsserver *nameserver.DNSServer) {
```

Comment:
Having this function, with a nearly identical, and long, signature to main.handleHTTP rather smells. I will have a go at refactoring that.